custom/ewsghana/reports/specific_reports/stock_status_report.py

```diff
@@ -382,6 +339,156 @@ class StockStatus(MultiReport):
+                    location_type__administrative=False,
+                ).order_by('name').exclude(supply_point_id__isnull=True)
+        return supply_points
+
+    def unique_products(self, locations, all=False):
+        if self.report_config['products'] and not all:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True),
+            ).filter(pk__in=self.report_config['products']).exclude(is_archived=True)
+        elif self.report_config['program'] and not all:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True),
+                program_id=self.report_config['program']
+            ).exclude(is_archived=True)
+        else:
+            return SQLProduct.objects.filter(
+                pk__in=locations.values_list('_products', flat=True)
+            ).exclude(is_archived=True)
+
+    def stockouts_data(self):
+        supply_points = self.get_supply_points.values_list('supply_point_id', flat=True)
+
+        if not supply_points:
+            return {}
+
+        unique_products = self.unique_products(self.get_supply_points, all=True)
+        transactions = StockTransaction.objects.filter(
+            type='stockonhand',
+            case_id__in=list(supply_points),
+            report__domain=self.report_config['domain'],
+            report__date__lte=self.report_config['enddate'],
+            report__date__gte=self.report_config['startdate'],
+            stock_on_hand=0,
+            product_id__in=list(unique_products.values_list('product_id', flat=True)),
+        ).distinct('case_id', 'product_id').order_by('case_id', 'product_id', '-report__date').values_list(
+            'case_id', 'product_id'
+        )
+        stockouts = defaultdict(set)
+
+        for (case_id, product_id) in transactions:
+            stockouts[case_id].add(product_id)
+
+        return {
+            'stockouts': stockouts,
+            'unique_products': unique_products,
+            'stockout_table_supply_points': supply_points
+        }
+
+    def data(self):
+        locations = self.report_location.get_descendants()
+        supply_points = locations.values_list('supply_point_id', flat=True)
+
+        if not supply_points:
+            return {}
+
+        unique_products = self.unique_products(locations, all=True)
+        transactions = StockTransaction.objects.filter(
+            type='stockonhand',
+            case_id__in=list(supply_points),
+            report__domain=self.report_config['domain'],
+            report__date__lte=self.report_config['enddate'],
+            report__date__gte=self.report_config['startdate'],
+            product_id__in=list(unique_products.values_list('product_id', flat=True)),
+        ).distinct('case_id', 'product_id').order_by('case_id', 'product_id', '-report__date').values_list(
+            'case_id', 'product_id', 'report__date', 'stock_on_hand'
+        )
```

Comment:
This query is complex enough that I'd pull it out to it's own method or function, especially as it's a duplicate.